File: macos/Sources/FloatingLyricCore/Playback/PlayheadClock.swift
import Foundation

/// Extrapolates the playback position between polls. Pure: callers supply `now`.
public final class PlayheadClock: @unchecked Sendable {
    public static let seekThresholdMs = 1500

    private var anchorProgressMs: Int = 0
    private var anchorTime: TimeInterval?
    private var playing = false

    public init() {}

    public func anchor(progressMs: Int, isPlaying: Bool, now: TimeInterval) {
        anchorProgressMs = max(0, progressMs)
        anchorTime = now
        playing = isPlaying
    }

    public func positionMs(now: TimeInterval) -> Int {
        guard let anchorTime else { return 0 }
        guard playing else { return anchorProgressMs }
        let elapsedMs = Int(((now - anchorTime) * 1000).rounded())
        return max(0, anchorProgressMs + elapsedMs)
    }

    /// True when a freshly polled progress differs from the extrapolated
    /// position by more than `seekThresholdMs`.
    public func isSeek(polledProgressMs: Int, now: TimeInterval) -> Bool {
        guard anchorTime != nil else { return false }
        return abs(polledProgressMs - positionMs(now: now)) > Self.seekThresholdMs
    }
}

File: windows/src/FloatingLyric.Core/PlayheadClock.cs
namespace FloatingLyric.Core;

/// <summary>
/// Extrapolates the playback position between polls, so a 3-second poll can
/// drive a 10 Hz display. Pure: callers supply "now".
/// </summary>
public sealed class PlayheadClock
{
    public const int SeekThresholdMs = 1500;

    private int _anchorProgressMs;
    private double? _anchorTime;
    private bool _playing;

    public void Anchor(int progressMs, bool isPlaying, double now)
    {
        _anchorProgressMs = Math.Max(0, progressMs);
        _anchorTime = now;
        _playing = isPlaying;
    }

    public int PositionMs(double now)
    {
        if (_anchorTime is not { } anchorTime) return 0;
        if (!_playing) return _anchorProgressMs;
        var elapsedMs = (int)Math.Round((now - anchorTime) * 1000);
        return Math.Max(0, _anchorProgressMs + elapsedMs);
    }

    /// <summary>True when a fresh poll disagrees with the extrapolation by
    /// more than the threshold — that is a seek, not drift.</summary>
    public bool IsSeek(int polledProgressMs, double now) =>
        _anchorTime is not null &&
        Math.Abs(polledProgressMs - PositionMs(now)) > SeekThresholdMs;
}
